expression/integration_test.go

```diff
@@ -3169,6 +3172,66 @@ func (s *testIntegrationSuite) TestIssues(c *C) {
 	tk.MustQuery("select * from t where cast(a as binary)").Check(testkit.Rows("1"))
 }
 
+func (s *testIntegrationSuite) TestFilterExtractFromDNF(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	defer s.cleanEnv(c)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a int, b int, c int)")
+
+	tests := []struct {
+		exprStr   string
+		result    string
+		remained  string
+		resultStr string
+	}{
+		{
+			exprStr: "a = 1 or a = 1 or a = 1",
+			result:  "[eq(test.t.a, 1)]",
+		},
+		{
+			exprStr: "a = 1 or a = 1 or (a = 1 and b = 1)",
+			result:  "[eq(test.t.a, 1) eq(test.t.b, 1)]",
+		},
+		{
+			exprStr: "(a = 1 and a = 1) or a = 1 or b = 1",
+			result:  "[or(or(and(eq(test.t.a, 1), eq(test.t.a, 1)), eq(test.t.a, 1)), eq(test.t.b, 1))]",
+		},
+		{
+			exprStr: "(a = 1 and b = 2) or (a = 1 and b = 3) or (a = 1 and b = 4)",
+			result:  "[eq(test.t.a, 1) or(eq(test.t.b, 2), or(eq(test.t.b, 3), eq(test.t.b, 4)))]",
+		},
+		{
+			exprStr: "(a = 1 and b = 1) or (a = 1 and b = 1 and c = 1) or (a = 1 and b = 1 and c > 2 and c < 3)",
+			result:  "[eq(test.t.a, 1) eq(test.t.b, 1) or(eq(test.t.c, 1), and(gt(test.t.c, 2), lt(test.t.c, 3)))]",
```

Comment:
This is not correct too?

Reply:
The extracted part is `a = 1 and b = 1`, the remained part is `c = 1 or (c > 2 and c < 3)`.
So the result is `a = 1 and b = 1 and (c = 1 or (c > 2 and c < 3))`.